support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/web/flow/OneTimeTokenAccountSaveRegistrationAction.java

```diff
@@ -16,16 +20,39 @@
  * @author Misagh Moayyed
  * @since 5.0.0
  */
+@Slf4j
 @RequiredArgsConstructor
 public class OneTimeTokenAccountSaveRegistrationAction extends AbstractAction {
     private final OneTimeTokenCredentialRepository repository;
+    private final IGoogleAuthenticator googleAuthenticatorInstance;
 
     @Override
     protected Event doExecute(final RequestContext requestContext) {
+        val uid = WebUtils.getAuthentication(requestContext).getPrincipal().getId();
+        val credential = WebUtils.getCredential(requestContext, OneTimeTokenCredential.class);
+
+        if (!StringUtils.isNumeric(credential.getToken())) {
+            LOGGER.error("Non-numeric OTP token provided: [{}]", credential.getToken());
+            return error();
+        }
+
+        val token = Integer.parseInt(credential.getToken());
         val account = requestContext.getFlowScope().get("key", OneTimeTokenAccount.class);
 
-        val uid = WebUtils.getAuthentication(requestContext).getPrincipal().getId();
-        repository.save(uid, account.getSecretKey(), account.getValidationCode(), account.getScratchCodes());
-        return success();
+        LOGGER.debug("Attempting to validate OTP token [{}] with [{}]", token, account);
+        val isCodeValid = this.googleAuthenticatorInstance.authorize(account.getSecretKey(), token);
```

Comment:
According to their source code, this [may raise a runtime `GoogleAuthenticatorException`](https://github.com/wstrange/GoogleAuth/blob/master/src/main/java/com/warrenstrange/googleauth/IGoogleAuthenticator.java#L111) which is not currently being handled. Not really sure what should be done in that case.

Reply:
I'm not sure I've ever seen that happen in the same call in `GoogleAuthenticatorAuthenticationHandler`, which makes me think that just going for a 500 at that point might be the right response.

Reply:
Gotcha. Fair enough.